### PR TITLE
make bind throwing non-fatal

### PR DESCRIPTION
**Pull request by mafintosh**

### CODE REVIEW

File: lib/io.js [hunk 1]
```diff
@@ -208,22 +215,47 @@ module.exports = class IO {
...
+    } catch (err) {
+      this._binding = null
+      throw err
+    }
+  }
+
   bind() {
     if (this._binding) return this._binding
-    this._binding = this._bindSockets()
+    this._binding = this._bind()
```
Comment (marcus-pousette-hp) [inline lib/io.js]: we would need binding overnship to called here like

```suggestion
    const binding = this._bindSockets().catch((err) => {
    if (this._binding === binding) this._binding = null
    throw err
  })

  this._binding = binding
```

else  concurrent bind and resume can flake 


```ts
test('failed bind does not clear a concurrent resume', async function (t) {
  const a = createDHT({ anyPort: false })
  t.teardown(() => a.destroy())
  await a.fullyBootstrapped()
  await a.suspend()

  const bindSockets = a.io._bindSockets
  let failures = 2
  a.io._bindSockets = function () {
    if (failures-- > 0) {
      return Promise.reject(new Error('simulated bind failure'))
    }
    return bindSockets.call(this)
  }

  await t.exception(a.resume(), /simulated bind failure/)

  // Start both before awaiting either.
  const binding = a.bind()
  const resuming = a.resume()
  await t.exception(binding, /simulated bind failure/)
  await resuming

  await t.execution(a.bind(), 'bind reuses the resumed sockets')
})
```

File: index.js [hunk 2]
```diff
@@ -159,8 +160,15 @@ class DHT extends EventEmitter {
...
 
   async suspend({ log = noop } = {}) {
+    // Let an in-flight resume finish first so the final state is suspended
+    try {
+      if (this.resuming) await this.resuming
+    } catch {}
     log('Suspending waiting for io bind...')
-    await this.io.bind()
+    // A failed bind must not prevent suspending, _clear tolerates missing sockets
+    try {
```
Comment (marcus-pousette-hp) [inline index.js]: We might need another  

```ts
if (this.suspended || this.destroyed) return
```

Edge case is a failed resume and then a suspend


```ts
test('suspend after failed resume keeps sockets closed', async function (t) {
  const a = createDHT({ ephemeral: false, firewalled: false })
  t.teardown(() => a.destroy())

  await a.fullyBootstrapped()
  await a.suspend()

  const bindSockets = a.io._bindSockets
  a.io._bindSockets = async function () {
    throw new Error('EADDRNOTAVAIL (simulated)')
  }

  await t.exception(a.resume(), /EADDRNOTAVAIL/)
  a.io._bindSockets = bindSockets

  await a.suspend()

  t.ok(a.io.serverSocket.closing, 'server socket stays closed')
  t.ok(a.io.clientSocket.closing, 'client socket stays closed')
})
```

**Review by marcus-pousette-hp — CHANGES_REQUESTED** [review]
I left two comments. There also seem to be an existing suspend/resume race bug unrelated to this PR that potentialyl could be fixed part of this PR, see below. A fix would be to wait for suspend cleanly until resume. 

```ts
test('overlapping suspend and resume preserve resumed I/O', async function (t) {
  const a = createDHT({ ephemeral: false, firewalled: false })
  await a.fullyBootstrapped()

  const watcher = a.io.networkInterfaces
  const socket = a.io.serverSocket
  const stream = a.udx.createStream(1)
  stream.on('error', () => {})

  t.teardown(async () => {
    stream.destroy()
    await a.destroy()
    await watcher.destroy()
  })

  // Keep the old socket from finishing close().
  stream.connect(socket, 2, socket.address().port, '127.0.0.1')

  const suspending = a.suspend()
  await new Promise(setImmediate)
  const resuming = a.resume()
  await new Promise(setImmediate)

  stream.destroy()
  await Promise.all([suspending, resuming])

  t.is(a.suspended, false)
  t.ok(a.io._drainInterval, 'resumed I/O keeps its drain timer')
})
```

Review was co-authored with AI
